lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4092,6 +4180,93 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     }
   }
 
+  void visitStructElementAddrInst(StructElementAddrInst *seai) {
+    auto loc = seai->getLoc();
+    auto &differentiationStrategies =
+        getDifferentiationTask()->getStructExtractDifferentiationStrategies();
+    auto strategy = differentiationStrategies.lookup(seai);
+    switch (strategy) {
+    case StructExtractDifferentiationStrategy::Inactive:
+      assert(!activityInfo.isActive(seai, synthesis.indices));
+      return;
+    case StructExtractDifferentiationStrategy::Fieldwise: {
+      // Compute adjoint as follows:
+      //   y = struct_element_addr x, #key
+      //   adj[x]#key' = struct_element_addr adj[y], #key'
+      //   adj[x]#key' += adj[y]
+      // where `#key'` is the field in the cotangent space corresponding to
+      // `#key`.
+      auto structTy = remapType(seai->getOperand()->getType()).getASTType();
+      auto cotangentVectorTy = structTy->getAutoDiffAssociatedVectorSpace(
+          AutoDiffAssociatedVectorSpaceKind::Cotangent,
+          LookUpConformanceInModule(getModule().getSwiftModule()))
+              ->getType()->getCanonicalType();
+      auto *cotangentVectorDecl =
+          cotangentVectorTy->getStructOrBoundGenericStruct();
+      assert(cotangentVectorDecl);
+      // Find the corresponding field in the cotangent space.
+      VarDecl *correspondingField = nullptr;
+      // If the cotangent space is the original struct, then field is the same.
+      if (cotangentVectorDecl == seai->getStructDecl())
+        correspondingField = seai->getField();
+      // Otherwise, look up the field by name.
+      else {
+        auto correspondingFieldLookup =
+        cotangentVectorDecl->lookupDirect(seai->getField()->getName());
+        assert(correspondingFieldLookup.size() == 1);
+        correspondingField = cast<VarDecl>(correspondingFieldLookup.front());
+      }
+      // Accumulate adjoint element buffer into adjoint buffer for original
+      // `struct_element_addr` operand.
+      auto adjStructElt =
+          builder.createStructElementAddr(loc,
+              getAdjointBuffer(seai->getOperand()), correspondingField);
+      auto adjElt = getAdjointBuffer(seai);
+      auto *readAccess = builder.createBeginAccess(
+          loc, adjElt, SILAccessKind::Read, SILAccessEnforcement::Static,
+          /*noNestedConflict*/ true, /*fromBuiltin*/ false);
+      addToAdjointBuffer(adjStructElt, readAccess);
+      builder.createEndAccess(loc, readAccess, /*aborted*/ false);
+      return;
+    }
+    case StructExtractDifferentiationStrategy::Getter: {
+      // Construct pullback arguments.
+      // Allocate buffer for pullback indirect result.
+      auto *pullbackResultBuf = builder.createAllocStack(
```

Comment:
I should probably change `pullbackResultBuf` to a `ValueWithCleanup` using `emitCleanup` and do cleanup before `builder.createDeallocStack(loc, pullbackResultBuf);`.

@rxwei: could you please confirm?

Reply:
There's no reason to create a cleanup, because the pullback's indirect result is strictly local to the visitor. Just `emitCleanup`.

Reply:
Thanks! Done in https://github.com/apple/swift/pull/22669/commits/9b80f4a605fcde05ad0ea8341892549982c82ba8.